src/lib/math/pcurves/pcurves_wrap.h

```diff
@@ -0,0 +1,236 @@
+      }
+
+      std::vector<uint8_t> serialize_scalar(const Scalar& scalar) const override {
+         return from_stash(scalar).serialize_to_vec();
+      }
+
+      std::optional<Scalar> deserialize_scalar(std::span<const uint8_t> bytes) const override {
+         if(auto scalar = C::Scalar::deserialize(bytes)) {
+            return stash(*scalar);
+         } else {
+            return {};
+         }
+      }
+
+      Scalar scalar_from_bits_with_trunc(std::span<const uint8_t> bytes) const override {
+         return stash(C::Scalar::from_bits_with_trunc(bytes));
+      }
+
+      std::optional<AffinePoint> deserialize_point(std::span<const uint8_t> bytes) const override {
+         if(auto pt = C::AffinePoint::deserialize(bytes)) {
+            return stash(*pt);
+         } else {
+            return {};
+         }
+      }
+
+      ProjectivePoint hash_to_curve(std::string_view hash,
+                                    std::span<const uint8_t> input,
+                                    std::span<const uint8_t> domain_sep,
+                                    bool random_oracle) const override {
+         if constexpr(C::ValidForSswuHash) {
+            return stash(hash_to_curve_sswu<C>(hash, random_oracle, input, domain_sep));
+         } else {
+            throw Not_Implemented("Hash to curve is not implemented for this curve");
+         }
+      }
+
+      Scalar scalar_add(const Scalar& a, const Scalar& b) const override {
+         return stash(from_stash(a) + from_stash(b));
+      }
+
+      Scalar scalar_sub(const Scalar& a, const Scalar& b) const override {
+         return stash(from_stash(a) - from_stash(b));
+      }
+
+      Scalar scalar_mul(const Scalar& a, const Scalar& b) const override {
+         return stash(from_stash(a) * from_stash(b));
+      }
+
+      Scalar scalar_square(const Scalar& s) const override { return stash(from_stash(s).square()); }
+
+      Scalar scalar_invert(const Scalar& s) const override { return stash(from_stash(s).invert()); }
+
+      Scalar scalar_negate(const Scalar& s) const override { return stash(from_stash(s).negate()); }
+
+      bool scalar_is_zero(const Scalar& s) const override { return from_stash(s).is_zero(); }
+
+      bool scalar_equal(const Scalar& a, const Scalar& b) const override { return from_stash(a) == from_stash(b); }
+
+      Scalar scalar_zero() const override { return stash(C::Scalar::zero()); }
+
+      Scalar scalar_one() const override { return stash(C::Scalar::one()); }
+
+      Scalar scalar_from_u32(uint32_t x) const override { return stash(C::Scalar::from_word(x)); }
+
+      Scalar random_scalar(RandomNumberGenerator& rng) const override { return stash(C::Scalar::random(rng)); }
+
+      PrimeOrderCurveImpl() : m_mul_by_g(C::G) {}
+
+      static std::shared_ptr<const PrimeOrderCurve> instance() {
+         static auto g_curve = std::make_shared<const PrimeOrderCurveImpl<C>>();
+         return g_curve;
+      }
```

Comment:
There is no (mutable) state in these - just the precomputed base mul table, which we certainly want to share.

I had hoped the whole table could be `constexpr`/`constinit` (in principle it could be) but clearly compilers are not anywhere close to up to it - several couldn't even handle a single field inversion before bailing out.

Reply:
Mhm, that was my intuition as well. Yeah, I guess there's no way around the singleton then. Out of the blue: perhaps it would make sense to define the precomputed table itself a singleton (rather then the entire `PrimeOrderCurve`) to make this clearer?